File: trunk/MVP.Sample/MVP.Base/BaseUserControl/BaseControlCollection.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Web;

namespace MVP.Base.BaseUserControl
{
    public static class BaseControlCollection
    {
        public static Dictionary<string, Dictionary<string, BaseUserControl>> Controls
        {
            get
            {
                if (HttpContext.Current.Session["SS_BaseControlCollection_Key"] == null)
                {
                    HttpContext.Current.Session["SS_BaseControlCollection_Key"] = new Dictionary<string, Dictionary<string, BaseUserControl>>();
                }
                return (Dictionary<string, Dictionary<string, BaseUserControl>>)HttpContext.Current.Session["SS_BaseControlCollection_Key"];
            }
        }

        public static void Register(string parentKey, string key, BaseUserControl control)
        {
            if (!string.IsNullOrEmpty(parentKey))
            {
                lock (Controls)
                {
                    if (Controls.ContainsKey(parentKey))
                    {
                        if (Controls[parentKey] == null)
                        {
                            Controls[parentKey] = new Dictionary<string, BaseUserControl>();
                        }
                        if (Controls[parentKey].ContainsKey(key))
                        {
                            Controls[parentKey][key] = control;
                        }
                        else
                        {
                            Controls[parentKey].Add(key, control);
                        }
                    }
                    else
                    {
                        Controls.Add(parentKey, new Dictionary<string, BaseUserControl>());
                        Controls[parentKey].Add(key, control);
                    }
                }
            }
        }
        
    }
}

File: trunk/DotNetNuke/DNN.MVP.Base/BaseUserControl/BaseControlCollection.vb
Imports System.Web

Namespace BaseUserControl
    Public Module BaseControlCollection
        Public ReadOnly Property Controls As Dictionary(Of String, Dictionary(Of String, IBaseUserControl))
            Get
                If (HttpContext.Current.Session("SS_BaseControlCollection_Key") Is Nothing) Then
                    HttpContext.Current.Session("SS_BaseControlCollection_Key") = New Dictionary(Of String, Dictionary(Of String, IBaseUserControl))()
                End If
                Return CType(HttpContext.Current.Session("SS_BaseControlCollection_Key"), Dictionary(Of String, Dictionary(Of String, IBaseUserControl)))
            End Get
        End Property

        Public Sub Register(ByVal parentKey As String, ByVal key As String, ByVal control As IBaseUserControl)
            If Not String.IsNullOrEmpty(parentKey) Then
                SyncLock (Controls)
                    If Controls.ContainsKey(parentKey) Then
                        If (Controls(parentKey) Is Nothing) Then
                            Controls(parentKey) = New Dictionary(Of String, IBaseUserControl)
                        End If
                        If (Controls(parentKey).ContainsKey(key)) Then
                            Controls(parentKey)(key) = control
                        Else
                            Controls(parentKey).Add(key, control)
                        End If
                    Else
                        Controls.Add(parentKey, New Dictionary(Of String, IBaseUserControl)())
                        Controls(parentKey).Add(key, control)
                    End If
                End SyncLock
            End If
        End Sub
    End Module
End Namespace